hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/impl/listener/ClientNonSmartListenerService.java

```diff
@@ -23,100 +23,134 @@
-        try {
-            ClientInvocationFuture future = invocation.invoke();
-            String registrationId = serializationService.toObject(future.get());
-            handler.onListenerRegister();
-            Address address = future.getInvocation().getSendConnection().getRemoteEndpoint();
-            ClientEventRegistration registration = new ClientEventRegistration(registrationId,
-                    request.getCallId(), address);
-            registrations.put(registrationKey, registration);
-        } catch (Exception e) {
-            //if invocation cannot be done that means connection is broken and there is no need to add listener
-            EmptyStatement.ignore(e);
-        }
+        handler.beforeListenerRegister();
+
+        ClientRequest addRequest = registrationKey.getAddRequest();
+        BaseClientRemoveListenerRequest removeRequest = registrationKey.getRemoveRequest();
+
+        ClientInvocation invocation = new ClientInvocation(client, handler, addRequest);
+
+        ClientInvocationFuture future = invocation.invoke();
+        String registrationId = serializationService.toObject(future.get());
+        handler.onListenerRegister();
+
+        Address address = future.getInvocation().getSendConnection().getRemoteEndpoint();
+        return new ClientEventRegistration(registrationId,
+                addRequest.getCallId(), address, removeRequest);
+
     }
 
+
     @Override
-    public boolean deregisterListener(BaseClientRemoveListenerRequest request, String userRegistrationId) {
-        ClientEventRegistration registration = registrations.remove(new ClientRegistrationKey(userRegistrationId));
-        if (registration == null) {
-            return false;
-        }
-        removeEventHandler(registration.getCallId());
-        request.setRegistrationId(registration.getServerRegistrationId());
-        try {
-            Future future = new ClientInvocation(client, request, registration.getSubscriber()).invoke();
-            future.get();
-        } catch (Exception e) {
-            //if invocation cannot be done that means connection is broken and listener is already removed
-            EmptyStatement.ignore(e);
+    public boolean deregisterListener(String userRegistrationId) {
+        synchronized (listenerRegLock) {
+            ClientRegistrationKey key = new ClientRegistrationKey(userRegistrationId);
+            ClientEventRegistration registration = registrations.get(key);
+
+            if (registration == null) {
+                return false;
+            }
+
+            BaseClientRemoveListenerRequest removeRequest = registration.getRemoveRequest();
+            removeRequest.setRegistrationId(registration.getServerRegistrationId());
+            try {
+                Future future = new ClientInvocation(client, removeRequest).invoke();
+                future.get();
+                removeEventHandler(registration.getCallId());
+                registrations.remove(key);
```

Comment:
don't you need listenerRegLock synchronization at this point as performed in registerListener?